library/lib.ml

```diff
@@ -423,38 +423,40 @@ type section_data = {
   sec_entry : secentry list;
   sec_workl : Opaqueproof.work_list;
   sec_abstr : abstr_list;
+  sec_poly  : bool;
 }
 
-let empty_section_data = {
+let empty_section_data ~poly = {
   sec_entry = [];
   sec_workl = (Names.Cmap.empty,Names.Mindmap.empty);
   sec_abstr = (Names.Cmap.empty,Names.Mindmap.empty);
+  sec_poly  = poly;
 }
 
 let sectab =
   Summary.ref ([] : section_data list) ~name:"section-context"
 
-let add_section () =
-  sectab := empty_section_data :: !sectab
-
-let check_same_poly p vars =
-  let pred = function Context _ -> p = false | Variable {poly} -> p != poly in
-  if List.exists pred vars then
+let check_same_poly p sec =
+  if p != sec.sec_poly then
     user_err Pp.(str  "Cannot mix universe polymorphic and monomorphic declarations in sections.")
 
+let add_section ~poly () =
+  List.iter (fun tab -> check_same_poly poly tab) !sectab;
```

Comment:
We only need to check the head (if it exists) no?

Reply:
I guess so. In that case, we could even set the boolean in the global data structure instead.

Reply:
I am not changing this, because it'd be possible to allow polymorphic sections inside monomorphic one in a logically meaningful way. I might implement it that way in the kernel, so let's leave open the ability to do that here as well.

Reply:
+1 that’s a use case I have, defining a polymorphic thing, closing the section and using a specific instance